bin/writer.sunxi_raw

```diff
@@ -33,3 +33,4 @@ echo "Writing to ${device} from build at ${build}"
 
 dd if=/dev/zero of=$device bs=1k count=1024
 dd if=${build}/u-boot-sunxi-with-spl.bin of=$device bs=1024 seek=8
+sync
```

Comment:
I think this is hiding some other issue.

I can reproduce this error *if* some application other than dd has the disk device open. In that case, when dd closes the device, it's very fast; I assume because the buffers aren't flushed because there's a non-zero reference count on the device. Then, if I immediately pull out the disk I see those same errors. However, if nothing else has the disk open, dd's close() is very slow because the buffers are being flushed.

While forcing a sync appears to help here, since it forces all buffers to disk, I believe this is just hiding the problem that some other process has the disk open. What if that other process writes to the disk and corrupts the data that dd just wrote? Even a sync wouldn't solve that.

Can you check `lsof /dev/sdd`. `/proc/mounts`, etc. when the dd is running to see if anything else has the disk open? Also, try running dd under strace to see how quickly the close() system call returns; it should be slow, but if it's immediate it indicates the buffers aren't being synchronously flushed as expected.

Reply:
I cannot see anything else.

/proc/mounts has nothing (there is no partition table after all) and dmesg just shows one line:

[214287.385882] sd 8:0:0:0: [sdd] 15523840 512-byte logical blocks: (7.95 GB/7.40 GiB)

i.e. no mount points.

lsof shows the dd process if I make it run long enought, e.g. with:

time dd if=$device of=/dev/null bs=1M count=100

in the script (which runs for about 4 seconds):

sudo lsof -r1 /dev/sdd
=======
=======
=======
=======
=======
=======
=======
=======
=======
=======
=======
=======
=======
=======
COMMAND   PID          USER   FD   TYPE DEVICE   SIZE/OFF NODE NAME
dd      15046 gitlab-runner    0r   BLK   8,48 0t11534336  760 /dev/sdd
=======
COMMAND   PID          USER   FD   TYPE DEVICE   SIZE/OFF NODE NAME
dd      15046 gitlab-runner    0r   BLK   8,48 0t37748736  760 /dev/sdd
=======
COMMAND   PID          USER   FD   TYPE DEVICE   SIZE/OFF NODE NAME
dd      15046 gitlab-runner    0r   BLK   8,48 0t65011712  760 /dev/sdd
=======
COMMAND   PID          USER   FD   TYPE DEVICE   SIZE/OFF NODE NAME
dd      15046 gitlab-runner    0r   BLK   8,48 0t91226112  760 /dev/sdd
=======
=======
=======
=======
=======

So I can't see any other reason for what is going on here. I wonder if dd is synchronous with respect to its process, but the disk caching layer is separate? E.g. I can write to a file with 'cat' but it doesn't end up really written to disk until I user 'sync' or wait a bit.

Reply:
> I wonder if dd is synchronous with respect to its process, but the disk caching layer is
> separate? E.g. I can write to a file with 'cat' but it doesn't end up really written to disk
> until I user 'sync' or wait a bit.

Yes, that's absolutely true usually. write() synchronously copies data from user-space to the kernel, into the buffer cache. close() in most cases doesn't do much beyond adjust file metadata in the buffer cache. Actual writes to disk may happen much later. The exception is for filehandles to block devices; the code I linked to from comments when reviewing the previous version of this file implement the exception that close() on a block device synchronously pushes all dirty pages from the buffer cache to the device (perhaps only when the file open count reaches 0). I'm confused why that works for me, but doesn't seem to for you. I wonder if the SD card itself has some kind of buffering and needs power kept around after the last SD protocol write command until the SD card has flushed its own internal buffers? I would have hope the SD card's firmware made this synchronous too, but perhaps not. Perhaps the solution is:
a) Some kind of protocol between the kernel and SD to wait for all outstanding IO internal to the device to complete (hopefully this already exists and could be enabled via some existing sysctl or similar?)
b) A specific length of sleep after dd exits, with the length hopefully defined by SD protocol spec or perhaps individual SD card device spec.

Is there any way to inspect Linux's buffer cache to determine which of dd's call to close(), sync's execution, or just waiting for a few seconds, clear it out for the SD card?

Reply:
I have no indication that the SD card is caching anything. It's behaviour is generally assumed to be synchronous, although I suppose it is possible that it is not.

I think we should just go with the 'sync'. It doesn't work without it. I believe the kernel is caching blocks and writing them after dd exits. That makes perfect sense, since we are not using conv=fdatasync (since it syncs after every block). I have found that option to be very slow.

See also this:

https://unix.stackexchange.com/questions/312687/why-is-sync-so-important-when-making-a-bootable-linux-usb-stick

So again, I think we should just go with the 'sync'

Reply:
I'm still not convinced this is a good solution. Sure, it might *seem* to work right now, but there's nothing magical about "1" that guarantees anything. If something else does a lot of CPU- or disk-intensive work while this script is running, we could easily need "sleep 2" or "sleep 30". This will cause your testing to have intermittent failures. In fact, the required sleep time is unbounded in general, since there's no guarantee that sleeping will ever guarantee to fully flush the disk buffers. We really need to work out what's causing the synchronous flush to fail, and fix that directly, since that will guarantee that no sleep will ever be required.

Is there a way you can use usbmon on the SD card adapter to check whether any disk writes/... happen after dd has complete? That would at least determine whether the issue really is disk buffer flushing catching up during the sleep, or something else, e.g. that the SD card/switcher needs a bit more time before switching its IO to the other host.

Reply:
Are we talking about the same thing? This change adds a 'sync' command, not a sleep. The more I have dug into that the more I have convinced myself that this is necessary and sufficient.

Reply:
Oh yes, this is sync not sleep. I still don't think it's necessary or sufficient. It should not be necessary since the kernel does this itself when the device is closed; see the kernel code I pointed at in the previous pull request. I suspect it's working simply because sync takes some time to flush buffers to other storage devices in the system, so it acts like an unpredictably long delay/sleep, and/or something else is holding the SD card open and sync is the thing that ends up flushing the SD card's buffers rather than dd's close() call. What was it that you found that make you think sync is necessary and/or sufficient?

Reply:
It is necessary because I need either that or a sleep to make it work.

It is sufficient because I have never seen a failure with the sync in place.

Did you see the post above?

Also this:

https://www.gnu.org/software/coreutils/manual/html_node/sync-invocation.html#sync-invocation

Also see the 'direct' flag for dd. It seems to be there to bypass caches.

Linux caches block devices so the only way to ensure that the cache is flushed to disk is with sync. A delay is not enough as you say, because the machine might be busy.

Reply:
Out of curiosity, if you replace "sync" with "sync $device" does this still appear to help? If so, that's a good indication there's some reason it's really necessary. If not, it's a good indication that "sync" is simply operating as an unpredictable "sleep".

Reply:
I'm changing things a bit so I can't easily test that at the moment. I'll let you know when I find out.

Reply:
I tried

sync $device

When it work it works. I think it is good enough to be reliable. We can certainly try it and from the man page it looks good.

Sometimes I get:

+++ dd if=../current/orangepi_pc/u-boot-sunxi-with-spl.bin of=/dev/sdcard1 bs=1024 seek=8
465+1 records in
465+1 records out
476579 bytes (477 kB, 465 KiB) copied, 0.0657753 s, 7.2 MB/s
+++ sync /dev/sdcard1
sync: error syncing '/dev/sdcard1': Input/output error
Exit code: 1

I never get that with just 'sync'

But it is a bit confusing as I think it may be related to this...if I don't do the sync then I see errors in syslog like this:

Feb 15 08:13:26 kea kernel: [89062.544387] scsi_io_completion: 2 callbacks suppressed
Feb 15 08:13:26 kea kernel: [89062.544389] sd 10:0:0:0: [sdf] tag#0 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
Feb 15 08:13:26 kea kernel: [89062.544391] sd 10:0:0:0: [sdf] tag#0 Sense Key : Not Ready [current] 
Feb 15 08:13:26 kea kernel: [89062.544392] sd 10:0:0:0: [sdf] tag#0 Add. Sense: Medium not present
Feb 15 08:13:26 kea kernel: [89062.544393] sd 10:0:0:0: [sdf] tag#0 CDB: Read(10) 28 00 00 00 10 00 00 00 08 00

After that the device does not work properly until I physically replug it.  Once things are in a stable state, 'sync $device' works for me.

Reply:
So I did a bunch more experiments:

`lsof|grep sde` takes 10s of seconds, so it's easy to miss short-lived processes.

`lsof` doesn't show processes that have started to execute `close()`, which is where my dd processes take the bulk of their time (since most writes go just to the disk cache, and `close()` is the slow part where they actually get committed to disk. Again, this makes it easy to miss processes.

`lsof /dev/sde` doesn't show a process that has `/dev/sde1` open but not `/dev/sde`, even though it's a child device. Again, this makes it easy to miss processes.

`blktrace` seems to not show the process that's initiating IO but rather the process that just happens to be running when disk blocks are flushed to disk.

All this is to say that determining exactly what process has your disk device open in addition to your `dd` will be very complex.

If I run `lsof /dev/sde` periodically, I sometimes see that `systemd-udevd` opens my disk. Presumably it's probing the device to see what filesystems are on it, or because it's trying to check whether a removable device has been removed. Uggh.

Typically `dd`'s `write()`s are very fast, and `close()` very slow while buffers are flushed. If another process also has the same disk open when `close()` is called by `dd`, the call to `close()` returns immediately since it's not the last process with the device open, so the buffers don't get flushed. Instead, when the other process calls `close()`, buffers are flushed then. This is true even if the other process specifies `O_RDONLY` when opening the device. I verified all this by watching `strace` output, and running a Python script as "the other process", which waited for user input before calling `close()`.

Coupling the last two observations means that it's entirely possible this happens:

- `dd` opens device.
- `dd` writes a ton of data to the device.
- `systemd-udev` opens the device for whatever reason.
- `dd` closes the device. This returns immediately since `systemd-udev` has it open too. This means `dd` exits earlier than expected.
- `systemd-udev` closes the device, which triggers the slow flush.

All this is to say that yes it seems like the `sync` is necessary when running under a general purpose OS (rather than a very controlled embedded system) since we probably can't prevent `systemd-udev` from opening the device. A `sync` on the specific device would be best.

Reply:
Oh dear I hadn't thought of that but of course the OS does check the partitions at some point. That certainly seems likely to me.

So are you OK with the pull request as is, with the sync $device ?

Reply:
Yes, if that works it should be OK.